### PR TITLE
Fix `c3sx` gate definition (`c3sqrtx`->`c3sx`)

### LINKED ISSUE #7148: Quantum circuits with a C3SX gate fail to convert back from qasm
### Information

- **Qiskit Terra version**: 0.18.3
- **Python version**: 3.8.5
- **Operating system**: macOS-10.16-x86_64-i386-64bit

### What is the current behavior?

A circuit that contains a C3SX gate can be translated to qasm, but the generated qasm cannot be translated back to a quantum circuit.
The error stems from a mismatch in the name of this gate between qelib1.inc ("c3sqrtx"):
https://github.com/Qiskit/qiskit-terra/blob/6af71f9d1e249a3747772449f299160df4047b83/qiskit/qasm/libs/qelib1.inc#L242
and qiskit's python code ("c3sx"):
https://github.com/Qiskit/qiskit-terra/blob/cf4005ff9fdc672111a827582e6b7d8ae2683be5/qiskit/circuit/library/standard_gates/x.py#L446
 https://github.com/Qiskit/qiskit-terra/blob/6af71f9d1e249a3747772449f299160df4047b83/qiskit/circuit/quantumcircuit.py#L1563

### Steps to reproduce the problem

```python
import qiskit

qc = qiskit.QuantumCircuit(4)
gate = qiskit.circuit.library.C3SXGate()
qc.append(gate, qargs=qc.qubits)

qasm = qc.qasm()
qc_from_qasm = qiskit.QuantumCircuit.from_qasm_str(qasm)  # fails:
# QasmError(
# qiskit.qasm.exceptions.QasmError: "Cannot find gate definition for 'c3sx', line 4 file "
```

### What is the expected behavior?

I would like the gate names to concur and the qasm to be convertible to a quantum circuit.

### Suggested solutions

- Rename the gate in qelib1.inc to "c3sx".
- Change the python code to use "c3sqrtx".

(This issue probably relates to #4943.)

image

### PR DESCRIPTION
**Pull request by jond01**
- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.

### Summary
Fixes #7148.
Change all the occurrences of `c3sqrtx` to `c3sx` to allow a quantum circuit with a
`C3SXGate` to be converted to qasm and back to a quantum circuit from the qasm.
Modify `test_loading_all_qelib1_gates` to truly test all `qelib1.inc` gates (except for specific ones).

### Details and comments
During working on this PR, I encountered more issues:

- Duplicate definitions and names of gates: `rcccx`, `c3x`, `c4x`, etc.
- Initial fixes for the issues should be similar to the one here.

Unfortunately, I would not be able to fix them all now.

### CODE REVIEW

File: test/python/circuit/test_circuit_load_from_qasm.py [hunk 1]
```diff
@@ -13,8 +13,9 @@
 
 """Test cases for the circuit qasm_file and qasm_string method."""
 
-import os
+from pathlib import Path
```
Comment (jakelishman) [inline test/python/circuit/test_circuit_load_from_qasm.py:16]: In general it's better not to make these unrelated changes; there's nothing especially wrong with `os.path`, it's just a little more verbose, but changing it causes churn in unrelated parts of the codebase.

File: test/python/circuit/test_circuit_load_from_qasm.py [hunk 2]
```diff
@@ -29,18 +30,16 @@ class LoadFromQasmTest(QiskitTestCase):
     def setUp(self):
         super().setUp()
         self.qasm_file_name = "entangled_registers.qasm"
-        self.qasm_dir = os.path.join(
-            os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "qasm"
-        )
-        self.qasm_file_path = os.path.join(self.qasm_dir, self.qasm_file_name)
+        self.qasm_dir = (Path(__file__).parent.parent / "qasm").absolute()
```
Comment (jakelishman) [inline test/python/circuit/test_circuit_load_from_qasm.py:33]: If you are going to use `pathlib`, you probably want to use `Path.parents[1]` rather than `Path.parent.parent`.

File: test/python/circuit/test_circuit_load_from_qasm.py [hunk 3]
```diff
@@ -54,14 +53,37 @@ def test_qasm_file(self):
         q_circuit_2.measure(qr_b, cr_d)
         self.assertEqual(q_circuit, q_circuit_2)
 
+    def get_qelib1_gates(self):
+        """
+        Get all the gates from qelib1.inc library as a set of strings
+        """
+        qelib1_path = (Path(qiskit.qasm.__file__).parent / "libs/qelib1.inc").absolute()
```
Comment (jakelishman) [inline test/python/circuit/test_circuit_load_from_qasm.py:60]: You should do `Path / "libs" / "qelib1.inc"` if you want to be safe; the form here implies that `/` is the file separator, which is the exact problem that both `os.path` and `pathlib` are designed to abstract, since it's not universally true.

File: test/python/circuit/test_circuit_load_from_qasm.py [hunk 4]
```diff
@@ -108,9 +130,28 @@ def test_loading_all_qelib1_gates(self):
...
+        ref_circuit.append(C3XGate(), [0, 1, 2, 3])
+        ref_circuit.append(C3SXGate(), [0, 1, 2, 3])
+        ref_circuit.append(C4XGate(), [0, 1, 2, 3, 4])
+
+        ref_circuit.measure_all()
+
+        # check that ref_circuit contains all the operations from qasm_circuit
+        ref_ops = ref_circuit.count_ops().keys()
+        qasm_ops = qasm_circuit.count_ops().keys()
+        self.assertTrue(qasm_ops <= ref_ops, msg=qasm_ops - ref_ops)
```
Comment (jakelishman) [inline test/python/circuit/test_circuit_load_from_qasm.py:145]: `self.assertLessEqual(qasm_ops, ref_ops)` should produce clearer logging output here (and below), but shouldn't they have exactly equal sets?

File: test/python/circuit/test_circuit_load_from_qasm.py [hunk 5]
```diff
@@ -108,9 +130,28 @@ def test_loading_all_qelib1_gates(self):
...
+
+        # check that all qelib1.inc gates are in ref_circuit
+        qelib1_gates = self.get_qelib1_gates()
+        qelib1_gates -= {"u0"}  # u0 is not supported outside of qelib1
+        qelib1_gates -= {"rc3x"}  # rc3x is defined as rcccx in RC3XGate class
+        qelib1_gates -= {"c3x"}  # c3x is defined as mcx in C3XGate class
+        qelib1_gates -= {"c4x"}  # c4x is defined as mcx in C4XGate class
+
+        self.assertTrue(qelib1_gates <= ref_ops, msg=qelib1_gates - ref_ops)
```
Comment (jakelishman) [inline test/python/circuit/test_circuit_load_from_qasm.py:154]: I'm not 100% sure what the purpose of this test is.  At any rate, `qelib1.inc` should be fixed, and not subject to additions and removals (we weren't great at that in the past), so I think it may not be necessary.

File: test/python/qasm/all_gates.qasm [hunk 6]
```diff
@@ -1,7 +1,13 @@
 OPENQASM 2.0;
 include "qelib1.inc";
-qreg q[3];
-creg c[3];
+
+gate rcccx_dg q0,q1,q2,q3 { u2(-2*pi,pi) q3; u1(pi/4) q3; cx q2,q3; u1(-pi/4) q3; u2(-2*pi,pi) q3; u1(pi/4) q3; cx q1,q3; u1(-pi/4) q3; cx q0,q3; u1(pi/4) q3; cx q1,q3; u1(-pi/4) q3; cx q0,q3; u2(-2*pi,pi) q3; u1(pi/4) q3; cx q2,q3; u1(-pi/4) q3; u2(-2*pi,pi) q3; }
+gate rcccx q0,q1,q2,q3 { u2(0,pi) q3; u1(pi/4) q3; cx q2,q3; u1(-pi/4) q3; u2(0,pi) q3; cx q0,q3; u1(pi/4) q3; cx q1,q3; u1(-pi/4) q3; cx q0,q3; u1(pi/4) q3; cx q1,q3; u1(-pi/4) q3; u2(0,pi) q3; u1(pi/4) q3; cx q2,q3; u1(-pi/4) q3; u2(0,pi) q3; }
+gate mcx q0,q1,q2,q3 { h q3; p(pi/8) q0; p(pi/8) q1; p(pi/8) q2; p(pi/8) q3; cx q0,q1; p(-pi/8) q1; cx q0,q1; cx q1,q2; p(-pi/8) q2; cx q0,q2; p(pi/8) q2; cx q1,q2; p(-pi/8) q2; cx q0,q2; cx q2,q3; p(-pi/8) q3; cx q1,q3; p(pi/8) q3; cx q2,q3; p(-pi/8) q3; cx q0,q3; p(pi/8) q3; cx q2,q3; p(-pi/8) q3; cx q1,q3; p(pi/8) q3; cx q2,q3; p(-pi/8) q3; cx q0,q3; h q3; }
+gate mcx_5272411680 q0,q1,q2,q3,q4 { h q4; cu1(-pi/2) q3,q4; h q4; rcccx q0,q1,q2,q3; h q4; cu1(pi/2) q3,q4; h q4; rcccx_dg q0,q1,q2,q3; c3sx q0,q1,q2,q4; }
```
Comment (jakelishman) [inline test/python/qasm/all_gates.qasm:7]: What are these extra gates doing, if the purpose of the test is to verify that things work with `qelib1.inc`?

If they are reasonable to keep, then we need to format them so they're a bit more legible.  Also, the last gate includes a gate `id`, which is generally not going to be reproducible from run to run, so it likely oughtn't to be there.

File: qiskit/qasm/libs/qelib1.inc [hunk 7]
```diff
@@ -260,7 +260,5 @@ gate c4x a,b,c,d,e
 {
     h e; cu1(pi/2) d,e; h e;
     c3x a,b,c,d;
-    h e; cu1(-pi/2) d,e; h e;
-    c3x a,b,c,d;
-    c3sqrtx a,b,c,e;
+    c3sx a,b,c,e;
```
Comment (jakelishman) [inline qiskit/qasm/libs/qelib1.inc:263]: Is this necessary?  I'm not keen on changing definitions unless it's actually wrong.

**Review by jakelishman — COMMENTED** [review]
Sorry nobody looked at this for a long time - it came in at quite a busy time, and fell through the cracks.

I'm not 100% sure that we're _safe_ to change the name of an operation in `qelib1.inc`; it may well have knock-on implications for how gates are serialised when being passed to the backends, so I'll need to check internally if it's actually safe for us to do that at all.

I left some other comments throughout - in general, this PR seems to be doing a few things, not just what the title says, so it's a little harder to review.  It's generally better to keep PRs to one issue, if possible, and not swap over library functions in working code, so it's easier to see what a change is doing.

**Comment by jond01** [comment]
> I'm not 100% sure that we're _safe_ to change the name of an operation in `qelib1.inc`; it may well have knock-on implications for how gates are serialised when being passed to the backends, so I'll need to check internally if it's actually safe for us to do that at all.

So I'm waiting for your decision before moving forward with this PR.
Please check the linked issue - there is an alternative solution.

**Comment by jakelishman** [comment]
I'm sorry that this PR went so long without a response.  Looking back, and I know you had a maintainer response on your issue before you started work, I don't think this approach was ever going to feel super comfortable, because it always involved changing some fixed data, either on the Qiskit or OpenQASM 2 side.  Both of these had the potential to break existing user code, for something that Qiskit _should_ be able to handle internally.

I've made #9183, which solves the problem in a much more conservative way.  I'm sorry that this PR got lost after the Christmas holidays last year, rather than me being able to work with you to merge it.
